File: trainer/src/wordle_evolution/ga/chromosome.py
"""The Chromosome (a bag of feature weights) plus species seeding helpers."""

from __future__ import annotations

import math
from dataclasses import dataclass, field

from ..wordle.types import FEATURE_ORDER

WEIGHT_MIN: float = -10.0
WEIGHT_MAX: float = 10.0


def clamp_weight(value: float) -> float:
    return max(WEIGHT_MIN, min(WEIGHT_MAX, float(value)))


def clamp_weights(weights: dict[str, float]) -> dict[str, float]:
    return {k: clamp_weight(v) for k, v in weights.items()}


@dataclass
class Chromosome:
    id: str
    weights: dict[str, float]
    mutation_rate: float
    generation_born: int
    species: str | None = None

    def __post_init__(self) -> None:
        # Guarantee every feature is present and clamped.
        self.weights = {f: clamp_weight(self.weights.get(f, 0.0)) for f in FEATURE_ORDER}
        self.mutation_rate = float(self.mutation_rate)

    def copy_with(self, **changes) -> "Chromosome":
        return Chromosome(
            id=changes.get("id", self.id),
            weights=changes.get("weights", dict(self.weights)),
            mutation_rate=changes.get("mutation_rate", self.mutation_rate),
            generation_born=changes.get("generation_born", self.generation_born),
            species=changes.get("species", self.species),
        )


# --------------------------------------------------------------------------- #
# Species seeding
# --------------------------------------------------------------------------- #
# Each champion species biases weights toward its namesake behaviour. Values are
# starting points; evolution refines them. All are within [-10, 10].
_SPECIES_TEMPLATES: dict[str, dict[str, float]] = {
    "Vowel Goblin": {
        "vowel_coverage_score": 8.0,
        "unique_letter_bonus": 5.0,
        "letter_frequency_score": 4.0,
        "duplicate_letter_penalty": -3.0,
        "known_absent_penalty": -4.0,
    },
    "Entropy Enjoyer": {
        "entropy_score": 9.0,
        "expected_remaining_penalty": -7.0,
        "candidate_bonus": 2.0,
        "known_absent_penalty": -3.0,
    },
    "Candidate Sniper": {
        "candidate_bonus": 8.0,
        "endgame_candidate_pressure": 7.0,
        "known_green_bonus": 5.0,
        "known_yellow_bonus": 4.0,
        "known_absent_penalty": -5.0,
    },
    "Frequency Gremlin": {
        "letter_frequency_score": 8.0,
        "positional_frequency_score": 7.0,
        "unique_letter_bonus": 3.0,
        "duplicate_letter_penalty": -4.0,
    },
    "Balanced Bot": {
        "entropy_score": 4.0,
        "candidate_bonus": 3.0,
        "letter_frequency_score": 3.0,
        "positional_frequency_score": 3.0,
        "known_green_bonus": 3.0,
        "known_yellow_bonus": 2.0,
        "known_absent_penalty": -3.0,
        "expected_remaining_penalty": -3.0,
        "vowel_coverage_score": 2.0,
        "unique_letter_bonus": 2.0,
        "duplicate_letter_penalty": -2.0,
        "endgame_candidate_pressure": 3.0,
    },
    "Mutation Gremlin": {
        "entropy_score": 3.0,
        "candidate_bonus": 3.0,
        "letter_frequency_score": 2.0,
        "positional_frequency_score": 2.0,
        "vowel_coverage_score": 2.0,
    },
}

SPECIES_NAMES: tuple[str, ...] = tuple(_SPECIES_TEMPLATES.keys())

# Species whose defining trait is a higher mutation rate.
_HIGH_MUTATION_SPECIES = {"Mutation Gremlin"}


def species_weights(species: str) -> dict[str, float]:
    """Full 12-feature weight dict for a named species (missing -> 0)."""
    template = _SPECIES_TEMPLATES.get(species, {})
    return {f: clamp_weight(template.get(f, 0.0)) for f in FEATURE_ORDER}


def species_mutation_rate(species: str, base: float) -> float:
    if species in _HIGH_MUTATION_SPECIES:
        return min(1.0, base * 2.5)
    return base


def random_weights(rng, low: float = WEIGHT_MIN, high: float = WEIGHT_MAX) -> dict[str, float]:
    """Uniform-random weights in ``[low, high]`` from a ``random.Random``-like rng."""
    return {f: clamp_weight(rng.uniform(low, high)) for f in FEATURE_ORDER}


def make_chromosome(
    chrom_id: str,
    weights: dict[str, float],
    mutation_rate: float,
    generation_born: int,
    species: str | None = None,
) -> Chromosome:
    return Chromosome(
        id=chrom_id,
        weights=clamp_weights(weights),
        mutation_rate=mutation_rate,
        generation_born=generation_born,
        species=species,
    )


# --------------------------------------------------------------------------- #
# Diversity / nickname
# --------------------------------------------------------------------------- #
def cosine_distance(a: dict[str, float], b: dict[str, float]) -> float:
    """1 - cosine similarity of two weight vectors (0 = identical direction)."""
    dot = 0.0
    na = 0.0
    nb = 0.0
    for f in FEATURE_ORDER:
        av = a.get(f, 0.0)
        bv = b.get(f, 0.0)
        dot += av * bv
        na += av * av
        nb += bv * bv
    if na == 0.0 or nb == 0.0:
        return 1.0
    sim = dot / (math.sqrt(na) * math.sqrt(nb))
    sim = max(-1.0, min(1.0, sim))
    return 1.0 - sim


# Dominant-weight -> playful nickname. Reuses the species vocabulary.
_NICKNAME_BY_FEATURE: dict[str, str] = {
    "candidate_bonus": "Candidate Sniper",
    "entropy_score": "Entropy Enjoyer",
    "expected_remaining_penalty": "The Pruner",
    "letter_frequency_score": "Frequency Gremlin",
    "positional_frequency_score": "Position Wizard",
    "unique_letter_bonus": "Caveman Guesser",
    "duplicate_letter_penalty": "Duplicate Hater",
    "vowel_coverage_score": "Vowel Goblin",
    "known_green_bonus": "The Chosen Bot",
    "known_yellow_bonus": "Yellow Chaser",
    "known_absent_penalty": "Grudge Keeper",
    "endgame_candidate_pressure": "Clutch Closer",
}


def nickname(chromosome: Chromosome) -> str:
    """Derive a playful label from the chromosome's dominant (max |weight|) feature."""
    if not chromosome.weights:
        return "Balanced Bot"
    dominant = max(FEATURE_ORDER, key=lambda f: abs(chromosome.weights.get(f, 0.0)))
    if abs(chromosome.weights.get(dominant, 0.0)) < 1e-9:
        return "Balanced Bot"
    return _NICKNAME_BY_FEATURE.get(dominant, "Balanced Bot")

File: apps/web/src/ga/chromosome.ts
// Chromosomes: the evolvable individuals.
//
// A chromosome is a bag of numeric weights (one per heuristic feature), a
// self-adaptive mutationRate, and some bookkeeping. Everything else in the GA
// operates on these.

import { Rng } from '../engine/seedRandom';
import {
  Chromosome,
  FEATURE_ORDER,
  FeatureName,
  FeatureWeights,
  weightsToArray,
} from '../engine/types';

/** Build a zeroed weights object. */
export function zeroWeights(): FeatureWeights {
  const w = {} as FeatureWeights;
  for (const f of FEATURE_ORDER) w[f] = 0;
  return w;
}

/** Random weights uniformly across the clamp range. */
export function randomWeights(rng: Rng, min: number, max: number): FeatureWeights {
  const w = {} as FeatureWeights;
  for (const f of FEATURE_ORDER) w[f] = rng.float(min, max);
  return w;
}

export function createRandomChromosome(
  rng: Rng,
  generationBorn: number,
  id: string,
  baseMutationRate: number,
  clampMin: number,
  clampMax: number,
): Chromosome {
  return {
    id,
    weights: randomWeights(rng, clampMin, clampMax),
    mutationRate: clamp(baseMutationRate * rng.float(0.6, 1.6), 0.01, 1),
    generationBorn,
  };
}

export function cloneChromosome(c: Chromosome, id?: string): Chromosome {
  return {
    id: id ?? c.id,
    weights: { ...c.weights },
    mutationRate: c.mutationRate,
    generationBorn: c.generationBorn,
  };
}

function clamp(x: number, lo: number, hi: number): number {
  return x < lo ? lo : x > hi ? hi : x;
}

// ---------------------------------------------------------------------------
// Seeded champion presets — hand-designed starting points so evolution has some
// good genes to recombine (and so the baseline comparison has real opponents).
// ---------------------------------------------------------------------------

export type PresetKey = 'entropy' | 'frequency' | 'candidate' | 'balanced';

export const CHAMPION_PRESETS: Record<PresetKey, { name: string; weights: FeatureWeights }> = {
  entropy: {
    name: 'Entropy enjoyer',
    weights: mk({
      candidateBonus: 3,
      entropyScore: 9,
      expectedRemainingPenalty: -6,
      letterFrequencyScore: 1,
      positionalFrequencyScore: 1,
      uniqueLetterBonus: 1,
      duplicateLetterPenalty: -2,
      vowelCoverageScore: 0.5,
      knownGreenBonus: 3,
      knownYellowBonus: 2,
      knownAbsentPenalty: -3,
      endgameCandidatePressure: 5,
    }),
  },
  frequency: {
    name: 'Frequency fiend',
    weights: mk({
      candidateBonus: 3,
      entropyScore: 0,
      expectedRemainingPenalty: 0,
      letterFrequencyScore: 7,
      positionalFrequencyScore: 8,
      uniqueLetterBonus: 3,
      duplicateLetterPenalty: -3,
      vowelCoverageScore: 3,
      knownGreenBonus: 4,
      knownYellowBonus: 3,
      knownAbsentPenalty: -4,
      endgameCandidatePressure: 6,
    }),
  },
  candidate: {
    name: 'Candidate sniper',
    weights: mk({
      candidateBonus: 8,
      entropyScore: 1,
      expectedRemainingPenalty: -1,
      letterFrequencyScore: 2,
      positionalFrequencyScore: 2,
      uniqueLetterBonus: 1,
      duplicateLetterPenalty: -1,
      vowelCoverageScore: 1,
      knownGreenBonus: 4,
      knownYellowBonus: 3,
      knownAbsentPenalty: -3,
      endgameCandidatePressure: 9,
    }),
  },
  balanced: {
    name: 'The balanced one',
    weights: mk({
      candidateBonus: 4,
      entropyScore: 5,
      expectedRemainingPenalty: -3,
      letterFrequencyScore: 4,
      positionalFrequencyScore: 4,
      uniqueLetterBonus: 2,
      duplicateLetterPenalty: -2,
      vowelCoverageScore: 2,
      knownGreenBonus: 4,
      knownYellowBonus: 3,
      knownAbsentPenalty: -3,
      endgameCandidatePressure: 6,
    }),
  },
};

function mk(partial: Partial<FeatureWeights>): FeatureWeights {
  return { ...zeroWeights(), ...partial };
}

/** Instantiate the seeded champions as chromosomes for a population. */
export function seededChampions(generationBorn: number, mutationRate: number): Chromosome[] {
  return (Object.keys(CHAMPION_PRESETS) as PresetKey[]).map((key) => ({
    id: `seed-${key}`,
    weights: { ...CHAMPION_PRESETS[key].weights },
    mutationRate,
    generationBorn,
  }));
}

/** Build a baseline chromosome for the comparison panel. */
export function presetChromosome(key: PresetKey): Chromosome {
  return {
    id: `baseline-${key}`,
    weights: { ...CHAMPION_PRESETS[key].weights },
    mutationRate: 0,
    generationBorn: 0,
  };
}

// ---------------------------------------------------------------------------
// Diversity metrics — cosine (direction) based, because the player uses argmax
// of a weighted sum and is therefore invariant to positive scaling. Euclidean
// distance would reward magnitude drift that changes no behaviour.
// ---------------------------------------------------------------------------

export function cosineDistance(a: FeatureWeights, b: FeatureWeights): number {
  const av = weightsToArray(a);
  const bv = weightsToArray(b);
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < av.length; i++) {
    dot += av[i] * bv[i];
    na += av[i] * av[i];
    nb += bv[i] * bv[i];
  }
  const denom = Math.sqrt(na) * Math.sqrt(nb);
  if (denom === 0) return na === nb ? 0 : 1;
  return 1 - dot / denom;
}

/** Mean cosine distance of the population to its (normalized) centroid. */
export function populationDiversity(pop: Chromosome[]): number {
  if (pop.length === 0) return 0;
  const n = FEATURE_ORDER.length;
  const units: Float64Array[] = [];
  const centroid = new Float64Array(n);

  for (const c of pop) {
    const v = weightsToArray(c.weights);
    let norm = 0;
    for (let i = 0; i < n; i++) norm += v[i] * v[i];
    norm = Math.sqrt(norm);
    const u = new Float64Array(n);
    if (norm > 0) for (let i = 0; i < n; i++) u[i] = v[i] / norm;
    units.push(u);
    for (let i = 0; i < n; i++) centroid[i] += u[i];
  }
  for (let i = 0; i < n; i++) centroid[i] /= pop.length;

  let cNorm = 0;
  for (let i = 0; i < n; i++) cNorm += centroid[i] * centroid[i];
  cNorm = Math.sqrt(cNorm);

  let sum = 0;
  for (const u of units) {
    if (cNorm === 0) {
      sum += 1;
      continue;
    }
    let dot = 0;
    for (let i = 0; i < n; i++) dot += u[i] * centroid[i];
    sum += 1 - dot / cNorm; // each u is already unit length
  }
  return sum / pop.length;
}

// ---------------------------------------------------------------------------
// Playful nicknames derived from a chromosome's dominant weights.
// ---------------------------------------------------------------------------

export function nickname(c: Chromosome): string {
  const w = c.weights;
  const top = dominantFeature(w);

  if (c.mutationRate > 0.4) return 'Mutation gremlin';
  if (magnitude(w) < 4) return 'Caveman guesser';

  switch (top) {
    case 'entropyScore':
    case 'expectedRemainingPenalty':
      return 'Entropy enjoyer';
    case 'candidateBonus':
    case 'endgameCandidatePressure':
      return 'Candidate sniper';
    case 'letterFrequencyScore':
    case 'positionalFrequencyScore':
      return 'Frequency fiend';
    case 'vowelCoverageScore':
      return 'Vowel goblin';
    case 'uniqueLetterBonus':
      return 'Letter spreader';
    default:
      return 'Curious bot';
  }
}

function dominantFeature(w: FeatureWeights): FeatureName {
  let best: FeatureName = FEATURE_ORDER[0];
  let bestVal = -Infinity;
  for (const f of FEATURE_ORDER) {
    const v = Math.abs(w[f]);
    if (v > bestVal) {
      bestVal = v;
      best = f;
    }
  }
  return best;
}

function magnitude(w: FeatureWeights): number {
  let s = 0;
  for (const f of FEATURE_ORDER) s += w[f] * w[f];
  return Math.sqrt(s);
}
